### ISSUE TITLE
`path.getTypeAnnotation()` throws

### ISSUE THREAD

**Issue by kangax** [babel/babel#5178]
Ran into this while fixing some Babili issues.

https://astexplorer.net/#/GbUDVxRZlD

### Input Code

```js
var a;
if (typeof a === "") {
  a;
}
```

### Expected Behavior
Should not throw

### Current Behavior
Throws "Invalid typeof value"

This is coming from `inferAnnotationFromBinaryExpression` which somehow produces `null`, passing it to `createTypeAnnotationBasedOnTypeof ` in flow.js which doesn't understand "null" as one of the types.

/cc'ing @jeffmo since this is possibly related to Flow.

**Comment by hzoo** [comment]
Oh thought it might of had to do with https://github.com/babel/babel/pull/3414 by @amasad but maybe not.

**Comment by jeffmo** [comment]
Hmm, AFAIK this isn't related to Flow other than syntax...but I'm not really familiar with this code.

Random guess: Does Babel try to infer types somewhere? If so, maybe it gets confused by the `""` typeof comparison?

**Comment by kangax** [comment]
Yep, AFAICS, Babel tries to infer annotation — https://github.com/babel/babel/blob/master/packages/babel-traverse/src/path/index.js#L38

This suite looks useful and explains things a little — https://github.com/babel/babel/blob/8c35b320d37a957b629e75f59a9f52292cfd6b8e/packages/babel-traverse/test/inference.js#L57-L175

Looks like this is just an inference bug. I'll fix it.
